### ISSUE TITLE
DatasetIterator created with BalancedPathFilter has no next element

### ISSUE THREAD

**Issue by tvanh** [deeplearning4j/deeplearning4j#6186]
import org.datavec.api.io.filters.BalancedPathFilter;
import org.datavec.api.io.filters.PathFilter;
import org.datavec.api.io.filters.RandomPathFilter;
import org.datavec.api.io.labels.ParentPathLabelGenerator;
import org.datavec.api.split.FileSplit;
import org.datavec.api.split.InputSplit;
import org.datavec.image.loader.NativeImageLoader;
import org.datavec.image.recordreader.ImageRecordReader;
import org.datavec.image.transform.ImageTransform;
import org.datavec.image.transform.PipelineImageTransform;
import org.deeplearning4j.datasets.datavec.RecordReaderDataSetIterator;
import org.deeplearning4j.earlystopping.EarlyStoppingConfiguration;
import org.deeplearning4j.earlystopping.EarlyStoppingModelSaver;
import org.deeplearning4j.earlystopping.EarlyStoppingResult;
import org.deeplearning4j.earlystopping.saver.LocalFileModelSaver;
import org.deeplearning4j.earlystopping.scorecalc.DataSetLossCalculator;
import org.deeplearning4j.earlystopping.termination.BestScoreEpochTerminationCondition;
import org.deeplearning4j.earlystopping.termination.MaxEpochsTerminationCondition;
import org.deeplearning4j.earlystopping.termination.MaxTimeIterationTerminationCondition;
import org.deeplearning4j.earlystopping.trainer.EarlyStoppingTrainer;
import org.deeplearning4j.eval.Evaluation;
import org.deeplearning4j.nn.conf.GradientNormalization;
import org.deeplearning4j.nn.conf.MultiLayerConfiguration;
import org.deeplearning4j.nn.conf.NeuralNetConfiguration;
import org.deeplearning4j.nn.conf.distribution.Distribution;
import org.deeplearning4j.nn.conf.distribution.GaussianDistribution;
import org.deeplearning4j.nn.conf.distribution.NormalDistribution;
import org.deeplearning4j.nn.conf.inputs.InputType;
import org.deeplearning4j.nn.conf.inputs.InvalidInputTypeException;
import org.deeplearning4j.nn.conf.layers.*;
import org.deeplearning4j.nn.multilayer.MultiLayerNetwork;
import org.deeplearning4j.nn.weights.WeightInit;
import org.deeplearning4j.optimize.api.TrainingListener;
import org.deeplearning4j.optimize.listeners.ScoreIterationListener;
import org.deeplearning4j.util.ModelSerializer;
import org.nd4j.linalg.activations.Activation;
import org.nd4j.linalg.dataset.api.iterator.DataSetIterator;
import org.nd4j.linalg.dataset.api.preprocessor.DataNormalization;
import org.nd4j.linalg.dataset.api.preprocessor.ImagePreProcessingScaler;
import org.nd4j.linalg.learning.config.Nesterovs;
import org.nd4j.linalg.lossfunctions.LossFunctions;
import org.nd4j.linalg.primitives.Pair;
import org.nd4j.linalg.schedule.ScheduleType;
import org.nd4j.linalg.schedule.StepSchedule;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.concurrent.TimeUnit;

import static java.lang.Math.toIntExact;

public class ImageClassifier {
    InputSplit trainData = null, testData  = null;
    DataSetIterator trainDSI = null, testDSI = null;
    int numLabels, numExamples, seed = 1*2*3*4, height, width, channels;
	
    protected int batchSize = 20;
	List<Pair<ImageTransform, Double>> imageTransforms = null;
	DataNormalization normalizer = null;
	protected PathFilter randomPathFilter, balancePathFilter;
	ImageRecordReader recordReader = null;
	
	public ImageClassifier setData(String path, int h, int w, int channel, double train_test_ratio){
		height = h;
		width = w;
		channels = channel;
		
	    Random rng = new Random(seed);
        ParentPathLabelGenerator labelMaker = new ParentPathLabelGenerator();
        File mainPath = new File(path);
        FileSplit fileSplit = new FileSplit(mainPath, NativeImageLoader.ALLOWED_FORMATS, rng);
        numExamples = toIntExact(fileSplit.length());
        numLabels = fileSplit.getRootDir().listFiles(File::isDirectory).length; //This only works if your root is clean: only label subdirs.
        //randomPathFilter = new RandomPathFilter(rng, NativeImageLoader.ALLOWED_FORMATS, 0);
        balancePathFilter = new BalancedPathFilter(rng, null, labelMaker, numExamples, numLabels, 0, 500);
        recordReader = new ImageRecordReader(h, w, channel, labelMaker);
        if (train_test_ratio < 1){
	        InputSplit[] inputSplit = fileSplit.sample(balancePathFilter, train_test_ratio, 1 - train_test_ratio);
	        trainData = inputSplit[0];
	        testData  = inputSplit[1];	        
        } else {
        	trainData = fileSplit.sample(balancePathFilter, 1)[0];
        }
        log.info("Load data completed....<NUMLABEL> classes, <NUMEXAMPLE> examples.".replace("<NUMLABEL>", Integer.toString(numLabels)).replace("<NUMEXAMPLE>", Integer.toString(numExamples)) );
        return this;
	}

	public ImageClassifier setBatchSize(int batchSize){
		this.batchSize = batchSize; 
		return this;
	}
	
	public ImageClassifier addPreprocessor(ImageTransform it, double ratio){
		if (imageTransforms == null){
			imageTransforms = new ArrayList<Pair<ImageTransform, Double>>();
		}
		imageTransforms.add(new Pair<>(it, ratio));
		return this;
	}
	
	public ImageClassifier normalize(DataNormalization normalizer){
		this.normalizer = normalizer;
		return this;
	}
	
	public void build() throws IOException{
        ImageTransform transform = null;
        if (imageTransforms != null){
        	transform =	new PipelineImageTransform(imageTransforms, true);
        }
        recordReader.initialize(trainData, transform);
        System.out.println("Train size: " + trainData.length());
        trainDSI = new RecordReaderDataSetIterator(recordReader, batchSize, 1, numLabels);
        if (normalizer != null){
            normalizer.fit(trainDSI);
            trainDSI.setPreProcessor(normalizer);
        }
        
        if (testData != null){
	        System.out.println("Test size: " + testData.length());
	        recordReader.initialize(testData, transform);
	        testDSI = new RecordReaderDataSetIterator(recordReader, batchSize, 1, numLabels);
	        if (normalizer != null){
		        normalizer.fit(testDSI);
		        testDSI.setPreProcessor(normalizer);
	        }
        }

	}
	
	public static enum CNNModel {
		Lenet, AlexNet;
	}
	
    protected static final Logger log = LoggerFactory.getLogger(ImageClassifier.class);

    protected static String modelType = "AlexNet"; // LeNet, AlexNet or Custom but you need to fill it out

    MultiLayerNetwork network = null;

    public void train(CNNModel model, TrainingListener tl, int epochs){
    	MultiLayerConfiguration config = null;
        switch (modelType) {
            case "LeNet":
            	config = lenetModel();
                break;
            case "AlexNet":
            	config = alexnetModel();
                break;
            default:
                throw new InvalidInputTypeException("Incorrect model provided.");
        }
        train(config, tl, epochs);
    }
    
    public void train(MultiLayerConfiguration config, TrainingListener tl, int epochs){
    	network = new MultiLayerNetwork(config);
        network.init();
        if (tl != null){
        	network.setListeners(tl);
        }
        network.fit(trainDSI, epochs);
    }
    
    public void train(MultiLayerConfiguration config, EarlyStoppingConfiguration<MultiLayerNetwork> esConfig){
    	EarlyStoppingTrainer trainer = new EarlyStoppingTrainer(esConfig, config, trainDSI);
        EarlyStoppingResult<MultiLayerNetwork> result = trainer.fit();
        network = result.getBestModel();
    }
    
    public void validate(){
    	if (network == null){
    		//LOG
    		return;
    	}
    	
    	if (testData == null){
    		//LOG
    		return;
    	}
    	
    	Evaluation eval = network.evaluate(testDSI);
        log.info(eval.stats(true));
    }
    
    public void save(File output) throws IOException{
    	if (!output.exists()){
    		if (!output.createNewFile()){
    			return;
    		}
    	}
        ModelSerializer.writeModel(network, output, true);
    }
    
    private ConvolutionLayer convInit(String name, int in, int out, int[] kernel, int[] stride, int[] pad, double bias) {
        return new ConvolutionLayer.Builder(kernel, stride, pad).name(name).nIn(in).nOut(out).biasInit(bias).build();
    }

    private ConvolutionLayer conv3x3(String name, int out, double bias) {
        return new ConvolutionLayer.Builder(new int[]{3,3}, new int[] {1,1}, new int[] {1,1}).name(name).nOut(out).biasInit(bias).build();
    }

    private ConvolutionLayer conv5x5(String name, int out, int[] stride, int[] pad, double bias) {
        return new ConvolutionLayer.Builder(new int[]{5,5}, stride, pad).name(name).nOut(out).biasInit(bias).build();
    }

    private SubsamplingLayer maxPool(String name,  int[] kernel) {
        return new SubsamplingLayer.Builder(kernel, new int[]{2,2}).name(name).build();
    }

    private DenseLayer fullyConnected(String name, int out, double bias, double dropOut, Distribution dist) {
        return new DenseLayer.Builder().name(name).nOut(out).biasInit(bias).dropOut(dropOut).dist(dist).build();
    }

    public MultiLayerConfiguration lenetModel() {
        /**
         * Revisde Lenet Model approach developed by ramgo2 achieves slightly above random
         * Reference: https://gist.github.com/ramgo2/833f12e92359a2da9e5c2fb6333351c5
         **/
        MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
            .seed(seed)
            .l2(0.005)
            .activation(Activation.RELU)
            .weightInit(WeightInit.XAVIER)
            .updater(new Nesterovs(1e-4, 1 - 1e-3))
            .list()
            .layer(0, convInit("cnn1", channels, 50 ,  new int[]{5, 5}, new int[]{1, 1}, new int[]{0, 0}, 0))
            .layer(1, maxPool("maxpool1", new int[]{2,2}))
            .layer(2, conv5x5("cnn2", 100, new int[]{5, 5}, new int[]{1, 1}, 0))
            .layer(3, maxPool("maxool2", new int[]{2,2}))
            .layer(4, new DenseLayer.Builder().nOut(500).build())
            .layer(5, new OutputLayer.Builder(LossFunctions.LossFunction.NEGATIVELOGLIKELIHOOD)
                .nOut(numLabels)
                .activation(Activation.SOFTMAX)
                .build())
            .backprop(true).pretrain(false)
            .setInputType(InputType.convolutional(height, width, channels))
            .build();

        return conf;

    }

    public MultiLayerConfiguration alexnetModel() {
        /**
         * AlexNet model interpretation based on the original paper ImageNet Classification with Deep Convolutional Neural Networks
         * and the imagenetExample code referenced.
         * http://papers.nips.cc/paper/4824-imagenet-classification-with-deep-convolutional-neural-networks.pdf
         **/

        double nonZeroBias = 1;
        double dropOut = 0.5;

        MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
            .seed(seed)
            .weightInit(WeightInit.DISTRIBUTION)
            .dist(new NormalDistribution(0.0, 0.01))
            .activation(Activation.RELU)
            .updater(new Nesterovs(new StepSchedule(ScheduleType.ITERATION, 1e-2, 0.1, 100000), 0.9))
            .biasUpdater(new Nesterovs(new StepSchedule(ScheduleType.ITERATION, 2e-2, 0.1, 100000), 0.9))
            .gradientNormalization(GradientNormalization.RenormalizeL2PerLayer) // normalize to prevent vanishing or exploding gradients
            .l2(5 * 1e-4)
            .list()
            .layer(0, convInit("cnn1", channels, 96, new int[]{11, 11}, new int[]{4, 4}, new int[]{3, 3}, 0))
            .layer(1, new LocalResponseNormalization.Builder().name("lrn1").build())
            .layer(2, maxPool("maxpool1", new int[]{3,3}))
            .layer(3, conv5x5("cnn2", 256, new int[] {1,1}, new int[] {2,2}, nonZeroBias))
            .layer(4, new LocalResponseNormalization.Builder().name("lrn2").build())
            .layer(5, maxPool("maxpool2", new int[]{3,3}))
            .layer(6,conv3x3("cnn3", 384, 0))
            .layer(7,conv3x3("cnn4", 384, nonZeroBias))
            .layer(8,conv3x3("cnn5", 256, nonZeroBias))
            .layer(9, maxPool("maxpool3", new int[]{3,3}))
            .layer(10, fullyConnected("ffn1", 4096, nonZeroBias, dropOut, new GaussianDistribution(0, 0.005)))
            .layer(11, fullyConnected("ffn2", 4096, nonZeroBias, dropOut, new GaussianDistribution(0, 0.005)))
            .layer(12, new OutputLayer.Builder(LossFunctions.LossFunction.NEGATIVELOGLIKELIHOOD)
                .name("output")
                .nOut(numLabels)
                .activation(Activation.SOFTMAX)
                .build())
            .backprop(true)
            .pretrain(false)
            .setInputType(InputType.convolutional(height, width, channels))
            .build();
        
        return conf;
    }

    public static MultiLayerNetwork customModel() {
        /**
         * Use this method to build your own custom model.
         **/
        return null;
    }
    
    public void testEarlyStop() throws IOException{
    	EarlyStoppingModelSaver<MultiLayerNetwork> saver = new LocalFileModelSaver("D:/data/zalo/musicgerne/model/");
        EarlyStoppingConfiguration<MultiLayerNetwork> esConf = new EarlyStoppingConfiguration.Builder<MultiLayerNetwork>()
                .epochTerminationConditions(new MaxEpochsTerminationCondition(100), new BestScoreEpochTerminationCondition(0.1)) 
                .evaluateEveryNEpochs(1)
                .iterationTerminationConditions(new MaxTimeIterationTerminationCondition(20, TimeUnit.MINUTES)) //Max of 20 minutes
                .scoreCalculator(new DataSetLossCalculator(testDSI, false)) 
                .modelSaver(saver)
                .build();
        		
        train(lenetModel(), esConf);
        validate();
        save(new File("D:/data/zalo/musicgerne/model/zalo_mfcc_model.bin"));
    }
    
    public void testTrainLenet() throws IOException{
    	train(lenetModel(), new ScoreIterationListener(1), 1); 
    }
    
    public static void main(String[] args) throws Exception {
    	String datapath = "C:/Users/anhtv/git/dl4j-examples/dl4j-examples/src/main/resources/animals";
    	int w = 100, h = 100, channel = 3;
    	ImageClassifier ic = new ImageClassifier();
    	double tt_ratio = 0.7;
    	ic.setData(datapath, h, w, channel, tt_ratio).setBatchSize(32).normalize(new ImagePreProcessingScaler(1, 0)).build();
    	ic.testEarlyStop();
    	//ic.testTrainLenet();
    }
}

cause
[1] The evalauation is NaN or 0
[2] Validate causes exeption
10:45:58.175 [main] INFO org.deeplearning4j.earlystopping.trainer.BaseEarlyStoppingTrainer - Starting early stopping training
10:46:00.003 [main] INFO org.deeplearning4j.earlystopping.trainer.BaseEarlyStoppingTrainer - Completed training epoch 0
10:46:00.003 [main] INFO org.deeplearning4j.earlystopping.trainer.BaseEarlyStoppingTrainer - Score at epoch 0: 0.0
10:46:01.235 [main] INFO org.deeplearning4j.earlystopping.trainer.BaseEarlyStoppingTrainer - Hit epoch termination condition at epoch 0. Details: BestScoreEpochTerminationCondition(0.1)
Exception in thread "main" java.util.NoSuchElementException: No next element
	at org.datavec.api.util.files.FileFromPathIterator.next(FileFromPathIterator.java:46)
	at org.datavec.api.util.files.FileFromPathIterator.next(FileFromPathIterator.java:33)
	at org.datavec.image.recordreader.BaseImageRecordReader.next(BaseImageRecordReader.java:234)
	at org.datavec.image.recordreader.BaseImageRecordReader.nextRecord(BaseImageRecordReader.java:496)
	at org.deeplearning4j.datasets.datavec.RecordReaderDataSetIterator.initializeUnderlying(RecordReaderDataSetIterator.java:218)
	at org.deeplearning4j.datasets.datavec.RecordReaderDataSetIterator.resetSupported(RecordReaderDataSetIterator.java:400)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.doEvaluationHelper(MultiLayerNetwork.java:3200)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.doEvaluation(MultiLayerNetwork.java:3192)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.evaluate(MultiLayerNetwork.java:3366)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.evaluate(MultiLayerNetwork.java:3284)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.evaluate(MultiLayerNetwork.java:3129)
	at anhtv.ai.dl.ImageClassifier.validate(ImageClassifier.java:206)
	at anhtv.ai.dl.ImageClassifier.testEarlyStop(ImageClassifier.java:331)
	at anhtv.ai.dl.ImageClassifier.main(ImageClassifier.java:350)

**Comment by saudet** [comment]
This is probably caused by how your data is laid out. Could you prepare an archive or a repository containing some data to reproduce this issue?

**Comment by tvanh** [comment]
No, nut you can reproduce this with your animal dataset (dl4j-examples/dl4j-examples/src/main/resources/animals).

I got the exception when testing this with the dataset above.

**Comment by agibsonccc** [comment]
Closing — this issue is over 7 years old with no recent activity. The codebase has undergone a major rewrite. If this is still relevant, please open a new issue with current reproduction steps.
